File: android/src/main/java/simpleconcept/dk/mobilepay_flutter/MobilePayFlutterPlugin.java
package simpleconcept.dk.mobilepay_flutter;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.NonNull;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import dk.mobilepay.sdk.Country;
import dk.mobilepay.sdk.MobilePay;
import dk.mobilepay.sdk.ResultCallback;
import dk.mobilepay.sdk.model.FailureResult;
import dk.mobilepay.sdk.model.Payment;
import dk.mobilepay.sdk.model.SuccessResult;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener;

/** MobilepayFlutterPlugin */
public class MobilePayFlutterPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware, ActivityResultListener {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private Context context;
  private Activity activity;
  int requestCode = 1337;

  public final Country getCountry(String countryCode) {
    if (countryCode.equals("fi")) {
      return Country.FINLAND;
    } else if(countryCode.equals("dk")) {
      return Country.DENMARK;
    } else {
      return Country.DENMARK;
    }
  }

  @Override
  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {

    Log.d("TESTING", "HERE1");

    if(requestCode == this.requestCode) {

      Log.d("TESTING", "HERE2");

      MobilePay.getInstance().handleResult(resultCode, data, new ResultCallback() {

        @Override
        public void onSuccess(SuccessResult result) {
          Log.d("TESTING", "SUCCESS");
          Map<String, Object> args  = new HashMap<String, Object>() {{
            put("requestCode", requestCode);
            put("amountWithdrawnFromCard", result.getAmountWithdrawnFromCard().doubleValue());
            put("orderId", result.getOrderId());
            put("signature", result.getSignature());
            put("transactionId", result.getTransactionId());
          }};
          channel.invokeMethod("mobilePaySuccess", args);
        }

        @Override
        public void onFailure(FailureResult result) {
          Log.d("TESTING", "FAILURE");
          Map<String, Object> args  = new HashMap<String, Object>() {{
            put("requestCode", requestCode);
            put("errorCode", result.getErrorCode());
            put("errorMessage", result.getErrorMessage());
          }};
          channel.invokeMethod("mobilePayFailure", args);
        }

        @Override
        public void onCancel(String orderId) {
          Log.d("TESTING", "CANCEL");
          Map<String, String> args  = new HashMap<String, String>() {{
            put("orderId", orderId);
          }};
          channel.invokeMethod("mobilePayCancel", args);
        }
      });
    }
    return true;
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "mobilepay_flutter");
    channel.setMethodCallHandler(this);
    context = flutterPluginBinding.getApplicationContext();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    Log.d("TESTING", call.method);
    switch (call.method) {
      case "initMobilePay":
        Log.d("TESTING", "INITING");
        String merchantId = call.argument("merchantId");
        String country = call.argument("country");
        MobilePay.getInstance().init(merchantId, getCountry(country));
        result.success(null);
        break;
      case "setRequestCode":
        requestCode = call.argument("requestCode");
        result.success(null);
        break;
      case "createPayment":
        double price = call.argument("productPrice");
        String orderId = call.argument("orderId");

        Payment payment = new Payment();
        payment.setProductPrice(BigDecimal.valueOf(price));
        payment.setOrderId(orderId);

        Intent paymentIntent = MobilePay.getInstance().createPaymentIntent(payment);

        activity.startActivityForResult(paymentIntent, requestCode);
        result.success(null);
        break;
      case "downloadMobilePay":
        Intent intent = MobilePay.getInstance().createDownloadMobilePayIntent(activity.getApplicationContext());
        activity.startActivity(intent);
        result.success(null);
        break;
      case "isMobilePayInstalled":
        Log.d("TESTING", "Got here...");
//        result.success(true);
        result.success(MobilePay.getInstance().isMobilePayInstalled(this.activity.getApplicationContext()));
        break;
      default: {
        result.notImplemented();
        break;
      }
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding activityBinding) {
    activity = activityBinding.getActivity();
    activityBinding.addActivityResultListener(this);
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
  }

  @Override
  public void onDetachedFromActivity() {
    activity = null;
  }
}

File: android/src/main/kotlin/simpleconcept/dk/mobilepay_flutter/MobilePayFlutterPlugin.kt
package simpleconcept.dk.mobilepay_flutter

import android.app.Activity
import android.content.Context
import dk.mobilepay.sdk.Country
import dk.mobilepay.sdk.MobilePay
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import dk.mobilepay.sdk.model.Payment
import java.math.BigDecimal
import android.content.Intent
import android.util.Log
import dk.mobilepay.sdk.model.FailureResult
import dk.mobilepay.sdk.model.SuccessResult
import dk.mobilepay.sdk.ResultCallback
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

class MobilePayFlutterPlugin {
    companion object : FlutterPlugin, MethodCallHandler, ActivityAware {

        private lateinit var channel : MethodChannel

        private lateinit var context: Context
        private lateinit var activity: Activity

        fun getCountry(countryCode: String): Country {
            when {
                countryCode == "fi" -> {
                    return Country.FINLAND
                }
                countryCode == "dk" -> {
                    return Country.DENMARK
                }
                else -> {
                    return Country.DENMARK
                }
            }
        }

        override fun onMethodCall(call: MethodCall, result: Result) {
            when {
                call.method == "initMobilePay" -> {
                    val merchantId = call.argument<String>("merchantId")
                    val country = call.argument<String>("country")
                    MobilePay.getInstance().init(merchantId!!, getCountry(country!!))
                    result.success(null)
                }
                call.method == "setRequestCode" -> {
                    val requestCode = call.argument<Int>("requestCode")

                    this.requestCode = requestCode!!
                    result.success(null)
                }
                call.method == "createPayment" -> {
                    val price = call.argument<Double>("productPrice")
                    val orderId = call.argument<String>("orderId")

                    val payment = Payment()
                    payment.productPrice = BigDecimal(price!!)
                    payment.orderId = orderId

                    val paymentIntent = MobilePay.getInstance().createPaymentIntent(payment)

                    activity.startActivityForResult(paymentIntent, requestCode)
                    result.success(null)
                }
                call.method == "downloadMobilePay" -> {
                    val intent = MobilePay.getInstance().createDownloadMobilePayIntent(context.applicationContext!!)
                    activity.startActivity(intent)
                    result.success(null)
                }
                call.method == "isMobilePayInstalled" -> {
                    val appContext = context.applicationContext!!
                    result.success(MobilePay.getInstance().isMobilePayInstalled(appContext))
                }
                else -> result.notImplemented()
            }
        }

        fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent): Boolean {
            if (requestCode == this.requestCode) {
                MobilePay.getInstance().handleResult(resultCode, data, object : ResultCallback {
                    override fun onSuccess(result: SuccessResult) {
                        val args = mapOf("requestCode" to requestCode,
                                "amountWithdrawnFromCard" to result.amountWithdrawnFromCard.toDouble(),
                                "orderId" to result.orderId,
                                "signature" to result.signature,
                                "transactionId" to result.transactionId)
                        channel.invokeMethod("mobilePaySuccess", args)
                    }

                    override fun onFailure(result: FailureResult) {
                        val args = mapOf("requestCode" to requestCode,
                                "orderId" to result.orderId,
                                "errorCode" to result.errorCode,
                                "errorMessage" to result.errorMessage
                        )
                        channel.invokeMethod("mobilePayFailure", args)
                    }

                    override fun onCancel(orderId: String) {
                        val args = mapOf("requestCode" to requestCode, "orderId" to orderId)
                        channel.invokeMethod("mobilePayCancel", args)
                    }
                })
            }
            return true
        }

        var requestCode: Int = 1337

        override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
            Log.d("***MsalFlutter***", "onAttachedToEngine")
            channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "mobilepay_flutter")
            channel.setMethodCallHandler(this);
            context = flutterPluginBinding.applicationContext
        }

        override fun onDetachedFromEngine(p0: FlutterPlugin.FlutterPluginBinding) {
            TODO("Not yet implemented")
        }

        override fun onAttachedToActivity(activityPluginBinding: ActivityPluginBinding) {
            activity = activityPluginBinding.activity;
        }

        override fun onDetachedFromActivityForConfigChanges() {
            TODO("Not yet implemented")
        }

        override fun onReattachedToActivityForConfigChanges(p0: ActivityPluginBinding) {
            TODO("Not yet implemented")
        }

        override fun onDetachedFromActivity() {
            TODO("Not yet implemented")
        }
    }
}
